Common/TableProducer/caloClusterProducer.cxx

```diff
@@ -261,35 +314,417 @@ struct caloClusterProducerTask {
+                             static_cast<o2::phos::ChannelType_t>(c.cellType()));
+      // TODO process MC info
+    }
+    // Set number of cells in last TrigRec
+    if (phosCellTRs.size() > 0) {
+      phosCellTRs.back().setNumberOfObjects(phosCells.size() - phosCellTRs.back().getFirstEntry());
+    }
+
+    // clusterize
+    if (isMC) {
+      o2::dataformats::MCTruthContainer<o2::phos::MCLabel> cellTruth;
+      o2::dataformats::MCTruthContainer<o2::phos::MCLabel> outputTruthCont;
+      clusterizerPHOS->processCells(phosCells, phosCellTRs, &cellTruth,
+                                    outputPHOSClusters, outputCluElements, outputPHOSClusterTrigRecs, outputTruthCont);
+    } else {
+      o2::dataformats::MCTruthContainer<o2::phos::MCLabel> dummyMC;
+      clusterizerPHOS->processCells(phosCells, phosCellTRs, nullptr,
+                                    outputPHOSClusters, outputCluElements, outputPHOSClusterTrigRecs, dummyMC);
+    }
+
+    // Find  CPV clusters corresponding to PHOS trigger records
+    std::vector<std::pair<float, float>> cpvMatchPoints[kCpvCells];
+    // Number of entries in each cell per TrigRecord
+    std::vector<trackTrigRec> cpvNMatchPoints;
+    cpvNMatchPoints.reserve(outputPHOSClusterTrigRecs.size());
+    cpvNMatchPoints.emplace_back();
+
+    int64_t curBC = cpvs.begin().bc_as<aod::BCsWithTimestamps>().globalBC();
+    cpvNMatchPoints.back().mTR = curBC;
+    for (int i = kCpvCells; i--;) {
+      cpvNMatchPoints.back().mStart[i] = 0;
+    }
+
+    for (const auto& cpvclu : cpvs) {
+      if (static_cast<int64_t>(cpvclu.bc_as<aod::BCsWithTimestamps>().globalBC()) != curBC) { // new BC
+        // mark last entry in previous range
+        for (int i = kCpvCells; i--;) {
+          cpvNMatchPoints.back().mEnd[i] = cpvMatchPoints[i].size();
+        }
+        curBC = cpvclu.bc_as<aod::BCsWithTimestamps>().globalBC();
+        cpvNMatchPoints.back() = cpvNMatchPoints.emplace_back();
+        cpvNMatchPoints.back().mTR = curBC;
+        for (int i = kCpvCells; i--;) {
+          cpvNMatchPoints.back().mStart[i] = cpvMatchPoints[i].size();
+        }
+      }
+      if (cpvclu.amplitude() < static_cast<std::vector<double>>(cpvMinE)[static_cast<int>(cpvclu.moduleNumber()) - 2]) {
+        continue;
+      }
+      int index = CpvMatchIndex(cpvclu.moduleNumber(), cpvclu.posX(), cpvclu.posZ());
+      cpvMatchPoints[index].emplace_back(cpvclu.posX(), cpvclu.posZ());
+    }
+    for (int i = kCpvCells; i--;) {
+      cpvNMatchPoints.back().mEnd[i] = cpvMatchPoints[i].size();
+    }
+    // same for tracks
+    std::vector<trackMatch> trackMatchPoints[kCpvCells]; // tracks hit in grid/cell in PHOS
+    // Number of entries in each cell per TrigRecord
+    std::vector<trackTrigRec> trackNMatchPoints;
+    trackNMatchPoints.reserve(outputPHOSClusterTrigRecs.size());
+    trackNMatchPoints.emplace_back();
+
+    curBC = tracks.begin().collision().bc_as<aod::BCsWithTimestamps>().globalBC();
```

Comment:
When using the sliceBy mentioned in my other comment, you could have a loop over collision that are in the current BC (should be one collision) and than use sliceBy to get the tracks associated with that collision and simply loop over these tracks.

Reply:
Yes, this is how EMCAL works, but my approach fits PHOS clusterization better.